test/e2e/apps/deployment.go

```diff
@@ -1705,3 +1709,215 @@ func testDeploymentSubresources(ctx context.Context, f *framework.Framework) {
 	framework.ExpectNoError(err, "Failed to get deployment resource: %v", err)
 	framework.ExpectEqual(*(deployment.Spec.Replicas), int32(4), "deployment should have 4 replicas")
 }
+
+func testPodTemplateHashCompatible(ctx context.Context, f *framework.Framework) {
+	ns := f.Namespace.Name
+	c := f.ClientSet
+
+	container := v1.Container{
+		Name:  "my-container",
+		Image: imageutils.GetPauseImageName(),
+	}
+
+	labelSelector := metav1.FormatLabelSelector(&metav1.LabelSelector{
+		MatchLabels: map[string]string{"test-name": "pod-template-hash-compatible"},
+	})
+
+	checkPodTemplateHash := func(ctx context.Context, podNum int, hasOldLabel, hasNewLabel bool) error {
+		pods, err := c.CoreV1().Pods(ns).List(context.Background(), metav1.ListOptions{
+			LabelSelector: labelSelector,
+		})
+		if err != nil {
+			return err
+		}
+		if len(pods.Items) != podNum {
+			return fmt.Errorf("wrong number of pod: expected %d, got %d", podNum, len(pods.Items))
+		}
+		for _, pod := range pods.Items {
+			oldLabel, hasOld := pod.Labels[appsv1.DefaultDeploymentUniqueLabelKey]
+			newLabel, hasNew := pod.Labels[appsv1.DeploymentUniqueLabelKey]
+			if hasNew != hasNewLabel {
+				return fmt.Errorf("expect pod to have %s label, got: %v, want: %v", appsv1.DeploymentUniqueLabelKey, hasNew, hasNewLabel)
+			}
+			if hasOld != hasOldLabel {
+				return fmt.Errorf("expect pod to have %s label, got: %v, want: %v", appsv1.DefaultDaemonSetUniqueLabelKey, hasOld, hasOldLabel)
+			}
+			if (hasNew && hasOld) && (oldLabel != newLabel) {
+				return fmt.Errorf("pod has inconsistent hash label")
+			}
+		}
+		return nil
+	}
+
+	checkRsPodTemplateHash := func(ctx context.Context, hasOldLabel, hasNewLabel, hasOldSelector, hasNewSelector bool) error {
+		rss, err := c.AppsV1().ReplicaSets(ns).List(ctx, metav1.ListOptions{
+			LabelSelector: labelSelector,
+		})
+		if err != nil {
+			return err
+		}
+		rsCnt := 0
+		for _, rs := range rss.Items {
+			if !(rs.Status.Replicas > 0 && rs.Status.ReadyReplicas == rs.Status.Replicas) {
+				continue
+			}
+			rsCnt++
+			if rsCnt > 1 {
+				return fmt.Errorf("expect only one ready rs, got more than 1")
+			}
+			if _, ok := rs.Spec.Selector.MatchLabels[appsv1.DefaultDeploymentUniqueLabelKey]; ok != hasOldSelector {
```

Comment:
nit: finding the RS first, and then doing the tests might look better.

Reply:
Not sure what you mean exactly, but I changed to get rs by deployment first and then check the rs, does this meet the requirements？

Reply:
yup, looks much better now. Please see also https://github.com/kubernetes/kubernetes/pull/116759#discussion_r1301715327